### PR TITLE
Fixes #246 : Forward plugin should detach problem listeners

### LINKED ISSUE #246: [ZF3] Forward plugin doesn't detach problem listeners anymore.
Looks like there was a regressions as part of a [v2 compatibility removable commit](https://github.com/zendframework/zend-mvc/commit/64c766d40c604e8a51523c1bc638453cd21acc5b#diff-a4e7671f65fba054e5c02177656ec841L212). The detachProblemListeners code builds the list of listeners to remove but doesn't actually remove them. For me, this manifests as my resulting ViewModel being double-rendered due to the InjectViewModelListener not actually being detached.

I will submit a PR to restore the block of code that actually detaches the listener.

### PR DESCRIPTION
**Pull request by nathanjosiah**
Fixes #246

### CODE REVIEW

File: src/Controller/Plugin/Forward.php [hunk 1]
```diff
@@ -194,6 +194,16 @@ protected function detachProblemListeners(SharedEvents $sharedEvents)
                         if (! is_object($currentCallback)) {
                             continue;
                         }
+
+                        foreach ($classArray as $class) {
+                            if ($currentCallback instanceof $class) {
+                                // Pass $currentEvent; when using zend-eventmanager v2,
```
Comment (froschdesign) [inline src/Controller/Plugin/Forward.php]: Please remove this comment, it was only relevant for v2.

File: test/Controller/Plugin/ForwardTest.php [hunk 2]
```diff
@@ -237,6 +238,35 @@ function ($e) {}
         $this->assertEquals(['content' => 'ZendTest\Mvc\Controller\TestAsset\ForwardController::testAction'], $result);
     }
 
+    public function testProblemListenersAreDetachedAndReattachedWhenPluginDispatchsRequestedController()
```
Comment (froschdesign) [inline test/Controller/Plugin/ForwardTest.php]: No assertion or expected exception?

**Review by froschdesign — CHANGES_REQUESTED** [review]

File: test/Controller/Plugin/ForwardTest.php (hunk 2 above)
Comment (nathanjosiah) [reply test/Controller/Plugin/ForwardTest.php]: @froschdesign I'm not sure what the ZF policy or practice is for this type of assertion but in this case, the assertion is implied in the mock method invocation expectation:
```php
$sharedEvents->expects($this->once())->method('detach')
```

If the method isn't invoked with the expected arguments, the test will fail with an exception. What is the preferred way of testing something like this? Maybe a test spy that monitors called methods and then the assertion can be the method was called?

Comment (nathanjosiah) [reply test/Controller/Plugin/ForwardTest.php]: @froschdesign Just bumping this, it's been 3 months and I'd love to start using ZF3. This issue is the only thing holding us back. 

I'm happy to make whatever change to the test. As my comment above says, my assertion is implied in the mock `once` invocation expectation. Would you prefer I change this to an explicit test spy? I'm not sure what the ZF test expectation would be for this use case.

**Comment by Xerkus** [comment]
Test failure is unrelated and fixed in https://github.com/zendframework/zend-mvc/pull/260

**Pull request merged**